PEP 696: Improve section "Using another TypeVarLike as the default" (#2777)
python/peps: pep-0696.rst

This PEP introduces the concept of type defaults for
``TypeVarLike``\ s (``TypeVar``, ``ParamSpec`` and ``TypeVarTuple``),

> which act as defaults for a type parameter when none is specified or

Reviewer: ```suggestion
which act as defaults for a type parameter when one is not specified or
```

Avoid potential confusion with `None`

 - `NumPy <https://github.com/numpy/numpy>`__ - the default for types
   like ``ndarray``'s ``dtype`` would be ``float64``. Currently it's
   ``Unknown`` or ``Any``.

>  - `TensorFlow <https://github.com/tensorflow/tensorflow>`__ - this

Reviewer: ```suggestion
 - `TensorFlow <https://github.com/tensorflow/tensorflow>`__ — this
```

Typographical nit, but this should be an em dash (you can also do `---` and Sphinx will replace it, if you don't remember the ALT code)

   class Foo(Generic[T, DefaultT]): ...   # Valid
   def bar(x: T, y: DefaultT): ...        # Valid
   class Foo(Generic[T]):
       class Bar(Generic[DefaultT]): ...  # Valid
   def outer(x: T):
       def inner(y: DefaultT): ...        # Valid

   StartT = TypeVar("StartT", default="StopT")  # Swapped defaults around from previous example
   StopT = TypeVar("StopT", default=int)
   class slice(Generic[StartT, StopT, StepT]): ...
                     # ^^^^^^ Invalid: ordering does not allow StopT to be bound
   def baz(x: DefaultT, y: T): ...
            # ^^^^^^^^ Invalid: ordering does not allow DefaultT to be bound

Bound Rules
~~~~~~~~~~~

``T2``'s bound must be a subtype of ``T1``'s bound.

.. code-block:: py

   T = TypeVar("T", bound=float)
   TypeVar("Ok", default=T, bound=int)        # Valid
   TypeVar("AlsoOk", default=T, bound=float)  # Valid
   TypeVar("Invalid", default=T, bound=str)   # Invalid: str is not a subtype of float

Constraint Rules
~~~~~~~~~~~~~~~~

Either ``T2``'s constraint must contain the bound of ``T1``, or the
constraints of ``T2`` must be a superset of the constraints of ``T1``.

.. code-block:: py

   T1 = TypeVar("T1", bound=int)

>    TypeVar("Ok", int, str, default=T1)                # Valid

Reviewer 2: This seems sketchy because `T1` may be solved to a subtype of `int`, but `Ok` can only be `int` or `str` exactly.

   TypeVar("Invalid", float, str, default=T1)         # Invalid: upper bound int is incompatible with constraints float or str

   T1 = TypeVar("T1", int, str)
   TypeVar("AlsoOk", int, str, bool, default=T1)      # Valid
   TypeVar("AlsoInvalid", bool, complex, default=T1)  # Invalid: {bool, complex} is not a superset of {int, str}

``TypeVarLike``\s as Parameters to Generics
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

``TypeVarLike``\ s are valid as parameters to generics inside of a
``default`` when the first parameter is in scope as determined by the
:ref:`previous section <scoping-rules>`.

.. code-block:: py

   T = TypeVar("T")
   ListDefaultT = TypeVar("ListDefaultT", default=list[T])

   class Bar(Generic[T, ListDefaultT]):

>       def __init__(self, x: T, y: ListDefaultT): ...

Reviewer 2: ```suggestion
       def __init__(self, x: T, y: ListDefaultT): ...
```

   DefaultTs = TypeVarTuple("DefaultTs", default=Unpack[tuple[str, int]])

   class Foo(Generic[*DefaultTs]): ...

   reveal_type(Foo())             # type is Foo[str, int]
   reveal_type(Foo[int, bool]())  # type is Foo[int, bool]

> Using Another ``TypeVarLike`` as ``default``

Author: Ok so coming back to this, I'm pretty confident that this is possible to implement (this sure isn't perfect and I had to disable some mypy sanity checks for things like default=T where T didn't use to be valid being passed) https://github.com/python/mypy/compare/master...Gobot1234:mypy:TypeVar-defaults.

```py
from typing import Callable, Generic, TypeVar
from typing_extensions import ParamSpec, TypeAlias, TypeVarTuple, Unpack, reveal_type

T = TypeVar("T", default=int)
T2 = TypeVar("T2", default=Callable[[T], int])

class Foo(Generic[T, T2]): ...

reveal_type(Foo)  # revealed type is type[__main__.Foo[T = builtins.int, T2 = (T = builtins.int) -> builtins.int]]
reveal_type(Foo[str])  # revealed type is type[__main__.Foo[builtins.str, T2 = (builtins.str) -> builtins.int]]
reveal_type(Foo[str, int])  # revealed type is type[__main__.Foo[builtins.str, int]]

PreSpecialised: TypeAlias = Foo[str]
reveal_type(PreSpecialised)    # revealed type is type[__main__.Foo[builtins.str, T2 = (builtins.str) -> builtins.int]]
reveal_type(PreSpecialised[int])  # This currently doesn't work but I don't think it will be hard to fix

P = ParamSpec("P", default=[int, str])
P2 = ParamSpec("P2", default=P)

class Bar(Generic[P, P2]): ...
def foo(fn: Callable[P, int]) -> bool: ...
reveal_type(foo)  # revealed type is def [P = [builtins.int, builtins.str]] (fn: (*P.args, **P.kwargs) -> builtins.int) -> builtins.bool
reveal_type(Bar)  # revealed type is type[__main__.Bar[P = [builtins.int, builtins.str], P2 = P = [builtins.int, builtins.str]]]
```
(I've slightly altered the representations mypy gives to make things a tiny bit clearer)
FYI this doesn't support TypeVarTuple defaults just cause mypy doesn't support PEP 646 yet, but enabling them should be about 20 to 30 lines of code

> Using Another ``TypeVarLike`` as ``default``

Author: I've just found out mypy does sort of support TypeVarTuples so I'll add support for that

> Using Another ``TypeVarLike`` as ``default``

Author: Ok this is actually a lot harder to support than I initially thought, I think I'm best waiting before trying to add this to mypy, there are a lot of gaps in the TypeVarTuple support. I can PR this feature in further down the road.

''''''''''''''''''''''''''''''''''''''''''''

This allows for a value to be used again when the constraints solver

> fails to solve a constraint for a type or the a type parameter to a
> generic is missed out but another is specified.

Reviewer: ```suggestion
fails to solve a constraint for a type, or the type parameter to a
generic is missing but another is specified.
```

There appears to be a typo here, and a missing comma. Also, I couldn't figure out precisely what is meant by "another" here—another _what_ is specified?

Author: Thanks, that reads better, I missed out type parameter so this now reads
> fails to solve a constraint for a type, or the type parameter to a
> generic is missing but a different type parameter is specified.

which I think is clearer

   class Foo(Generic[T, DefaultT]): ...   # Valid
   def bar(x: T, y: DefaultT): ...        # Valid
   class Foo(Generic[T]):
       class Bar(Generic[DefaultT]): ...  # Valid
   def outer(x: T):
       def inner(y: DefaultT): ...        # Valid

   StartT = TypeVar("StartT", default="StopT")  # Swapped defaults around from previous example
   StopT = TypeVar("StopT", default=int)
   class slice(Generic[StartT, StopT, StepT]): ...
                     # ^^^^^^ Invalid: ordering does not allow StopT to be bound
   def baz(x: DefaultT, y: T): ...
            # ^^^^^^^^ Invalid: ordering does not allow DefaultT to be bound

Bound Rules
~~~~~~~~~~~

``T2``'s bound must be a subtype of ``T1``'s bound.

.. code:: py

   T = TypeVar("T", bound=float)
   TypeVar("Ok", default=T, bound=int)        # Valid
   TypeVar("AlsoOk", default=T, bound=float)  # Valid
   TypeVar("Invalid", default=T, bound=str)   # Invalid: str is not a subtype of float

Constraint Rules
~~~~~~~~~~~~~~~~

``T2``'s constraint must either contain the bound of ``T1`` or the
constraints of ``T2`` must be a superset of the constraints of ``T1``.

.. code:: py

   T1 = TypeVar("T1", bound=int)

>    TypeVar("Ok", int, str default=T1)                 # Valid

Reviewer 2: ```suggestion
   TypeVar("Ok", int, str, default=T1)                 # Valid
```

   TypeVar("Invalid", float, str, default=T1)         # Invalid: upper bound int is incompatible with constraints float or str

   T1 = TypeVar("T1", int, str)
   TypeVar("AlsoOk", int, str, bool, default=T1)      # Valid
   TypeVar("AlsoInvalid", bool, complex, default=T1)  # Invalid: {bool, complex} is not a superset of {int, str}

``TypeVarLike``\ s As Parameters To Generics
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

``TypeVarLike``\ s are valid as parameters to generics inside of a
``default`` when the first parameter is in scope as determined by the
previous section.

.. code:: py

   T = TypeVar("T")
   ListDefaultT = TypeVar("ListDefaultT", default=list[T])

>    def foo(x: ListDefaultT): ...

Reviewer 2: I'm having a hard time understanding what this would mean.

Author: You're right this doesn't make sense and should be an error, whoops

   StopT = TypeVar("StopT", default=StartT)
   StepT = TypeVar("StepT", default=int | None)

   class slice(Generic[StartT, StopT, StepT]): ...

   reveal_type(slice())                        # type is slice[int, int, int | None]
   reveal_type(slice[str]())                   # type is slice[str, str, int | None]
   reveal_type(slice[str, bool, timedelta]())  # type is slice[str, bool, timedelta]

When using a ``TypeVarLike`` as the default to another ``TypeVarLike``.
Where ``T1`` is the default for ``T2`` the following rules apply.

Scoping Rules
~~~~~~~~~~~~~

> ``T1`` must be used before ``T2`` in the parameter list of the generic

Reviewer: ```suggestion
``T1`` must be used before ``T2`` in the parameter list of the generic,
```

Help make the intended grouping more clear, given the double or

or be bound in an outer class or function scope.

.. code:: py

   DefaultT = TypeVar("DefaultT", default=T)

   class Foo(Generic[T, DefaultT]): ...   # Valid
   def bar(x: T, y: DefaultT): ...        # Valid
   class Foo(Generic[T]):
       class Bar(Generic[DefaultT]): ...  # Valid
   def outer(x: T):
       def inner(y: DefaultT): ...        # Valid

   StartT = TypeVar("StartT", default="StopT")  # Swapped defaults around from previous example
   StopT = TypeVar("StopT", default=int)
   class slice(Generic[StartT, StopT, StepT]): ...
                     # ^^^^^^ Invalid: ordering does not allow StopT to be bound
   def baz(x: DefaultT, y: T): ...
            # ^^^^^^^^ Invalid: ordering does not allow DefaultT to be bound

Bound Rules
~~~~~~~~~~~

``T2``'s bound must be a subtype of ``T1``'s bound.

> .. code:: py

Reviewer: FYI, ``code-block`` is the modern Sphinx directive and should be preferred; ``code`` is the legacy docutils version with different behavior in some areas.

Author: Good to know

   T = TypeVar("T", bound=float)
   TypeVar("Ok", default=T, bound=int)        # Valid
   TypeVar("AlsoOk", default=T, bound=float)  # Valid
   TypeVar("Invalid", default=T, bound=str)   # Invalid: str is not a subtype of float

Constraint Rules
~~~~~~~~~~~~~~~~

> ``T2``'s constraint must either contain the bound of ``T1`` or the

Reviewer: ```suggestion
Either ``T2``'s constraint must contain the bound of ``T1``, or the
```

A comma here helps disambiguate the intended grouping (assuming I've parsed the intent correctly as "(T2's constraint must contain the bound of T2) or (the constrains of T2 must be a superset of the constraints of T2)"

constraints of ``T2`` must be a superset of the constraints of ``T1``.

.. code:: py

   T1 = TypeVar("T1", bound=int)
   TypeVar("Ok", int, str, default=T1)                # Valid
   TypeVar("Invalid", float, str, default=T1)         # Invalid: upper bound int is incompatible with constraints float or str

   T1 = TypeVar("T1", int, str)
   TypeVar("AlsoOk", int, str, bool, default=T1)      # Valid
   TypeVar("AlsoInvalid", bool, complex, default=T1)  # Invalid: {bool, complex} is not a superset of {int, str}

> ``TypeVarLike``\ s As Parameters To Generics
> ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Reviewer: Another typographical nit, but since this was specifically changed to use title case for h3+ headings (most PEPs I've seen tend to use sentence case for those headings, but up to you—just be consistent), you don't capitalize articles, short conjunctions, etc. in title case. Also, FYI, for this an the rest, you can just escape the `s` directly rather than the space before the `s`.

```suggestion
``TypeVarLike``\s as Parameters to Generics
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

``TypeVarLike``\ s are valid as parameters to generics inside of a
``default`` when the first parameter is in scope as determined by the

> previous section.

Reviewer: Can you `:ref:` the "previous section" you're talking about here, for clarity?

.. code:: py

   T = TypeVar("T")
   ListDefaultT = TypeVar("ListDefaultT", default=list[T])

   class Bar(Generic[T, ListDefaultT]):
      def __init__(self, x: T, y: ListDefaultT): ...

   reveal_type(Bar[int])             # type is Bar[int, list[int]]
   reveal_type(Bar[int, list[str]])  # type is Bar[int, list[str]]
   reveal_type(Bar[int, str])        # type is Bar[int, str]

Specialisation Rules
~~~~~~~~~~~~~~~~~~~~

``TypeVarLike``\ s currently cannot be further subscripted. This might

> change if Higher Kinded TypeVars are implemented.

Reviewer: Assuming "Higher Kinded TypeVars" is indeed a proper name of a concept or proposal, as implied by the case, could you link what that is here?

Subclasses of ``Generic``\ s with ``TypeVarLike``\ s that have defaults
behave similarly to ``Generic`` ``TypeAlias``\ es.

.. code-block:: py

   class SubclassMe(Generic[T, DefaultStrT]): ...

   class Bar(SubclassMe[int, DefaultStrT]): ...
   reveal_type(Bar())        # type is Bar[str]
   reveal_type(Bar[bool]())  # type is Bar[bool]

   class Foo(SubclassMe[float]): ...

>    reveal_type(Foo())  # type is <subclass of SubclassMe[int, float]>

Reviewer 2: Shouldn't this be `SubclassMe[float, str]`? Also, the reveal_type() output would just be "type is Foo".

Author: Yeah you're correct on both points. The `<subclass of SubclassMe[int, str]>` is just to show `DefaultStrT` is constrained to `str` in `SubclassMe`. To avoid this I could actually use `DefaultStrT` as an attribute but I think this might be a little harder to follow. Do you have any other ideas as to how else this could be shown?

Rejected Alternatives
---------------------

> Allowing The ``TypeVarLike``\ s Defaults To Be Passed To ``type.__new__``'s ``**kwargs``
> ''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''

Reviewer: Whether the first "to" is capitalized depends on whether you follow AP that says yes for to in infinitives, or most of the other style guides (APA, MLA, Chicago, etc) that say no. I generally follow AP by default, but it is somewhat of an outlier here so its really up to you

```suggestion
Allowing the ``TypeVarLike``\s Defaults to Be Passed to ``type.__new__``'s ``**kwargs``
'''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''
```

or

```suggestion
Allowing the ``TypeVarLike``\s Defaults To Be Passed to ``type.__new__``'s ``**kwargs``
'''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''
```

   class Box(Generic[T = int]):
       value: T | None = None

> Allowing Non-defaults To Follow Defaults

Reviewer: AP is as written, otherwise

```suggestion
Allowing Non-Defaults to Follow Defaults
```
